### PR TITLE
cli: set LESS=FRX when running $PAGER

### PR DESCRIPTION
**Pull request by ddevault**
Fixes https://github.com/martinvonz/jj/issues/3502

First time using jj to send a pull request, so some patience is appreciated.

I will not be signing the CLA, so do with this as you may. I certify the following in its stead:

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

### CODE REVIEW

File: cli/src/ui.rs [hunk 1]
```diff
@@ -132,7 +132,9 @@ impl UiOutput {
     fn new_paged(pager_cmd: &CommandNameAndArgs) -> io::Result<UiOutput> {
         let mut cmd = pager_cmd.to_command();
         tracing::info!(?cmd, "spawning pager");
-        let mut child = cmd.stdin(Stdio::piped()).spawn()?;
+        // LESS=FRX causes escape sequences to render correctly when PAGER=less
+        // It's harmless when using other pagers
+        let mut child = cmd.env("LESS", "FRX").stdin(Stdio::piped()).spawn()?;
```
Comment (yuja) [inline cli/src/ui.rs:137]: This was discussed recently, but I think `$LESS` should be set only if it wasn't sent by the current environment nor config.
https://github.com/martinvonz/jj/pull/3657#discussion_r1833657841

Comment (ddevault) [reply cli/src/ui.rs:137]: Oh, that's fair. I'll update the patch accordingly.

**Comment by supermarin** [comment]
:+1:

**Comment by martinvonz** [comment]
I asked the Google lawyers how to deal with the lack of a signed CLA. They said we're not allowed to even look at the code changes (which is what I expected, so I didn't look before asking either). So I'm afraid we'll have to close this PR, and I suppose we're not technically allowed to accept a PR from @yuja or anyone else who has looked at the code either. I have my open PR #3657 to address the same problem. I just haven't found time to get back to it, and it sounds like the consensus was to do it in a simpler way than I had. I appreciate the effort, @ddevault. I'm sorry we can't accept it.

**Comment by thoughtpolice** [comment]
FWIW, it's unclear the above fix even meets what would be known as the "threshold of originality" for copyright in the US, if it's just setting some launch variable; I strongly suspect it (and many other obvious 1-to-lets-say-10-line fixes) would fall under this threshold. There has been some precedent for this in other projects, like `mold`, which used this before its move to MIT. But I guess it's a matter of if OSPO thinks this is relevant or not in the mean time. (In the future I suspect we'd still require sign offs for all non-trivial changes, but this kind of change is one where we wouldn't really require it.)

**Comment by martinvonz** [comment]
> and I suppose we're not technically allowed to accept a PR from @yuja or anyone else who has looked at the code either.

Let me reinforce that "I suppose" just in case: I am not a lawyer

**Comment by ddevault** [comment]
> I asked the Google lawyers how to deal with the lack of a signed CLA. They said we're not allowed to even look at the code changes (which is what I expected, so I didn't look before asking either). So I'm afraid we'll have to close this PR, and I suppose we're not technically allowed to accept a PR from @yuja or anyone else who has looked at the code either. I have my open PR #3657 to address the same problem. I just haven't found time to get back to it, and it sounds like the consensus was to do it in a simpler way than I had. I appreciate the effort, @ddevault. I'm sorry we can't accept it.

I have some follow-up questions. How do you deal with other Apache 2.0-licensed code you find in the wild? Are you allowed to read the code of your dependencies even if the authors haven't signed the Google CLA? Is running "cargo add" subject to the same scrutiny? Are you allowed to reuse or integrate code from sources that haven't signed the CLA on your own authority, like a StackOverflow answer (CC-BY-SA)?

**Comment by ilyagr** [comment]
The official answers to these questions are public, in https://opensource.google/documentation/reference/thirdparty#all_non-google_or_open_source_code_and_data_goes_into_third_party.

Less relevantly, I think https://opensource.google/documentation/reference/patching is worth reading to see how corporations think of their employees contributing to open source. My IANAL understanding most corporations make people sign agreements that all of their work or ideas even vaguely related to the company's interests (so, for Google, any software work) is owned by the company.

There are companies that have stricter rules than Google. AFAIK, Apple refuses to license their employee's work to open-source projects with few exceptions (like contributing to Swift). So, Apple employees are essentially forbidden from contributing to open source, even in their free time. I am not sure whether they are even allowed to file a bug report.

**Comment by ddevault** [comment]
> The official answers to (a version of) these questions are public, in https://opensource.google/documentation/reference/thirdparty#all_non-google_or_open_source_code_and_data_goes_into_third_party.

This applies to Google's internal monorepo, not personal projects of Googlers.

**Comment by ddevault** [comment]
> So I'm afraid we'll have to close this PR, and I suppose we're not technically allowed to accept a PR from @yuja or anyone else who has looked at the code either

Seeing as @yuja is the second-most prolific contributor to jj after yourself it may be in the project's best interests for you to have a more two-sided conversation about policy with the lawyers rather than asking them to hand down an edict.

**Comment by ilyagr** [comment]
> > The official answers to (a version of) these questions are public, in https://opensource.google/documentation/reference/thirdparty#all_non-google_or_open_source_code_and_data_goes_into_third_party.
> 
> This applies to Google's internal monorepo, not personal projects of Googlers.

Ah, perhaps I found a better link: https://opensource.google/documentation/reference/releasing/preparing#third-party_components and https://opensource.google/documentation/reference/releasing/preparing#third-party_components.

This applies to "first-party Google projects" like Chromium. I don't know and don't want to speculate on whether `jj` counts as that, and whether all of those rules are necessary for it, but they'd probably be sufficient (for a Google employee asking themselves most of your questions from https://github.com/martinvonz/jj/pull/4834#issuecomment-2472722649). I interpret the latter link as saying "`cargo add` is fine".

https://opensource.google/documentation/reference/releasing/contributions is relevant as well.

I still recommend https://opensource.google/documentation/reference/patching  as follow-up reading for what kind of procedures Google likes.

**Comment by ilyagr** [comment]
I understand that this is basically the opposite of your point: you are actually interested in whether we could get away with doing less or negotiate better terms. Still, my point is that the questions in https://github.com/martinvonz/jj/pull/4834#issuecomment-2472722649 have answers, usually written by lawyers.

**Pull request closed without merging**